File: python/texture.py
import math
import random
import copy
from PIL import Image

from vector3 import *

class Texture(object):
	def value(self, u, v, p):
		return Vector3(0,0,0)

class Color(Texture):
	def __init__(self, color):
		self.color = color

	def value(self, u, v, p):
		return self.color

class Checker(Texture):
	def __init__(self, odd, even):
		self.odd = odd
		self.even = even

	def value(self, u, v, p):
		a = 10
		sines = math.sin(a*p.x)*math.sin(a*p.y)*math.sin(a*p.z)
		if (sines < 0):
			return self.odd.value(u,v,p)
		else:
			return self.even.value(u,v,p)

def trilinear_interp(c, u, v, w):
	accum = 0
	for i in range(2):
		for j in range(2):
			for k in range(2):
				kernel = (i*u + (1-i)*(1-u)) * (j*v+ (1-j)*(1-v)) * (k*w + (1-k)*(1-w))
				accum += c[i][j][k] * kernel
	return accum

def perlin_interp(c, u, v, w):
	uu = u*u*(3-2*u)
	vv = v*v*(3-2*v)
	ww = w*w*(3-2*w)
	accum = 0
	for i in range(2):
		for j in range(2):
			for k in range(2):
				weight_v = Vector3(u-i, v-j, w-k)
				kernel = (i*uu + (1-i)*(1-uu)) * (j*vv+ (1-j)*(1-vv)) * (k*ww+(1-k)*(1-ww))
				accum += c[i][j][k].dot(weight_v) * kernel
	return accum

class Perlin(object):
	# TODO: MAKE ALL OF THIS STATIC
	def __init__(self):
		self.randvec = []
		self.perm_x = []
		self.perm_y = []
		self.perm_z = []
		r = lambda: (random.random()*2) - 1
		for i in range(256):
			self.randvec.append(Vector3(r(), r(), r()).unit())
			self.perm_x.append(i)
			self.perm_y.append(i)
			self.perm_z.append(i)
		random.shuffle(self.perm_x)
		random.shuffle(self.perm_y)
		random.shuffle(self.perm_z)

	def noise(self, pt):
		u = pt.x - math.floor(pt.x)
		v = pt.y - math.floor(pt.y)
		w = pt.z - math.floor(pt.z)
		i = math.floor(pt.x)
		j = math.floor(pt.y)
		k = math.floor(pt.z)
		c = [[[0,0], [0,0]], [[0,0], [0,0]]]
		for di in range(2):
			for dj in range(2):
				for dk in range(2):
					idx = self.perm_x[(i+di) & 255] ^ self.perm_y[(j+dj) & 255] ^ self.perm_z[(k+dk) & 255]
					c[di][dj][dk] = self.randvec[idx]
		return perlin_interp(c, u, v, w)

	def turb(self, pt, depth):
		accum = 0.0
		temp_p = copy.copy(pt)
		weight = 1.0
		for i in range(depth):
			accum += weight * self.noise(temp_p)
			weight *= 0.5
			temp_p *= 2
		return abs(accum)

class NoiseTexture(Texture):
	def __init__(self, scale):
		self.perlin = Perlin()
		self.scale = scale

	def value(self, u, v, p):
		# Straight noise
		# Range is -0.7..0.7, found empirically.. ;/
		# offset = (1.0 / 0.7) * 0.5
		# noise = (self.perlin.noise(self.scale * p) * offset) + offset
		# Straight Turbulence
		# noise = self.perlin.turb(self.scale * p, 7)
		# Sin modulated by turb
		noise = 0.5 * (1 + math.sin(self.scale * p.x + 5 * self.perlin.turb(self.scale * p, 7)))
		return noise * one()

def clamp(val, minimum, maximum):
	return min(maximum, max(val, minimum))

class ImageTexture(Texture):
	def __init__(self, imageFilename):
		self.image = Image.open(imageFilename)

	def value(self, u, v, p):
		i = u*self.image.size[0]
		j = (1-v)*self.image.size[1]-0.001 # Hrmmmmm
		i = clamp(i, 0, self.image.size[0])
		j = clamp(j, 0, self.image.size[1])
		pixel = self.image.getpixel((i, j))
		ret = Vector3(pixel[0], pixel[1], pixel[2])
		ret /= 255.0
		return ret

File: rust/raytrace/src/texture.rs
extern crate image;

use std::rc::Rc;
use std::path::Path;
use self::image::{DynamicImage, GenericImageView, open, Pixel};

use vec3::Vec3;
use perlin::Perlin;

pub trait Texture {
  // result: attenuation, scatter
  fn value(&self, u: f64, v: f64, p: &Vec3) -> Vec3;
}

pub struct ConstantTexture {
  color: Vec3,
}

impl ConstantTexture {
  pub fn new(color: Vec3) -> ConstantTexture {
    ConstantTexture {
      color
    }
  }

  pub fn rc(color: Vec3) -> Rc<ConstantTexture> {
    Rc::new(ConstantTexture::new(color))
  }
}

impl Texture for ConstantTexture {
  fn value(&self, _u: f64, _v: f64, _p: &Vec3) -> Vec3 {
    self.color
  }
}

pub struct CheckerTexture {
  odd: Rc<Texture>,
  even: Rc<Texture>
}

impl CheckerTexture {
  pub fn new(odd: Rc<Texture>, even: Rc<Texture>) -> CheckerTexture {
    CheckerTexture {
      even,
      odd
    }
  }

  pub fn rc(odd: Rc<Texture>, even: Rc<Texture>) -> Rc<CheckerTexture> {
    Rc::new(CheckerTexture::new(odd, even))
  }
}

impl Texture for CheckerTexture {
  fn value(&self, u: f64, v: f64, p: &Vec3) -> Vec3 {
    let period_factor: f64 = 10.0;
    let px: f64 = p.x * period_factor;
    let py: f64 = p.y * period_factor;
    let pz: f64 = p.z * period_factor;
    let sines: f64 = px.sin() * py.sin() * pz.sin();
    if sines < 0.0 {
      self.odd.value(u, v, p)
    } else {
      self.even.value(u, v, p)
    }
  }
}

pub struct NoiseTexture {
  scale: f64,
  perlin: Perlin
}

impl NoiseTexture {
  pub fn new(scale: f64) -> NoiseTexture {
    NoiseTexture {
      scale,
      perlin: Perlin::new()
    }
  }

  pub fn rc(scale: f64) -> Rc<NoiseTexture> {
    Rc::new(NoiseTexture::new(scale))
  }
}

impl Texture for NoiseTexture {
  fn value(&self, _u: f64, _v: f64, p: &Vec3) -> Vec3 {
    // let offset = (1.0 / 0.8) * 0.5;
    // Straight noise, range is -0.8..0.8
    // Vec3::one() * ((self.perlin.noise(&(*p * self.scale)) * offset) + offset)
    // Striaght turb
    // Vec3::one() * self.perlin.turb(&(self.scale * *p), 7)
    // let noise = 0.5 * (1.0 + (self.scale * p.x + 5.0 * self.perlin.turb(&(self.scale * *p), 7)).sin());
    let noise = 0.5 * (1.0 + (self.scale * p.z + 5.0 * self.perlin.turb(&(self.scale * *p), 7)).sin());
    noise * Vec3::one()
  }
}

pub struct ImageTexture {
  image: DynamicImage
}

impl ImageTexture {
  pub fn new(filename: &Path) -> ImageTexture {
    ImageTexture {
      image: open(filename).unwrap()
    }
  }

  pub fn rc(filename: &Path) -> Rc<ImageTexture> {
    Rc::new(ImageTexture::new(filename))
  }
}

impl Texture for ImageTexture {
  fn value(&self, u: f64, v: f64, _p: &Vec3) -> Vec3 {
    let (width, height) = self.image.dimensions();
    let i = (u * width as f64).min(width as f64);
    let j = ((1.0 - v) * height as f64 - 0.001).min(height as f64);
    let pixel = self.image.get_pixel(i as u32, j as u32).to_rgb();
    let ret = Vec3::new(pixel[0] as f64, pixel[1] as f64, pixel[2] as f64);
    ret / 255.0
  }
}